modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/resolver/EvaluateAndTransformValueResolver.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.runtime.resolver;
+
+import static org.mule.util.Preconditions.checkArgument;
+import org.mule.api.MuleContext;
+import org.mule.api.MuleEvent;
+import org.mule.api.MuleException;
+import org.mule.api.transformer.MessageTransformer;
+import org.mule.api.transformer.Transformer;
+import org.mule.api.transformer.TransformerException;
+import org.mule.extension.introspection.DataType;
+import org.mule.transformer.types.DataTypeFactory;
+import org.mule.util.AttributeEvaluator;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * A {@link ValueResolver} which evaluates a MEL expressions and tries
+ * to ensure that the output is always of a certain type.
+ * <p/>
+ * If the MEL expression does not return a value of that type, then it
+ * tries to locate a {@link Transformer} which can do the transformation
+ * from the obtained type to the expected one.
+ * <p/>
+ * It resolves the expressions by making use of the {@link AttributeEvaluator}
+ * so that it's compatible with simple expressions and templates alike
+ *
+ * @param <T>
+ * @since 3.7.0
+ */
+public class EvaluateAndTransformValueResolver<T> implements ValueResolver<T>
```

Comment:
it only transforms if necessary while it always evaluate... if you don't like the name we can change it but the one you proposed I don't like either